From 9b3ecbda406c8363426f7af104704e84781b2370 Mon Sep 17 00:00:00 2001
From: Mateusz Poliwczak <mpoliwczak34@gmail.com>
Date: Tue, 4 Nov 2025 19:58:09 +0100
Subject: [PATCH 044/133] std.ArrayList: memset to undefined in
 shrinkRetainingCapacity and clearRetainingCapacity
X-Developer-Signature: v=1; a=openpgp-sha256; l=963; i=xtex@astrafall.org;
 h=from:subject; bh=ctw3AV4+TkMqzYH9U+8G1vFK02ZPKaoZIOlzMNmmD/w=;
 b=owGbwMvMwCW2U4Ij7wZL9ETG02pJDJn7cv/f0Jcp2Kj//+zWX2amDu9Vpmiv5Rb80nXqSQbbT
 L2HExcu6pjIwiDGxWAppshSZNjgzaqTzi+6rFwWZg4rE8gQaZEGBgYGBhYGvtzEvFIjHSM9U21D
 PUNDHSMdIwYuTgGYaoFqhv/eztd83A8851jKldVun1j4Ppx7Y6jESXsG1u8xHPHai/0ZGX6lzZ2
 r8GbnPr0MFoaSfT5yncUVYRZ3ft2deOHLpxCpbwwA
X-Developer-Key: i=xtex@astrafall.org; a=openpgp;
 fpr=7231804B052C670F15A6771DB918086ED8045B91

Fixes #25796

(cherry picked from commit 416bf1de47154ba781082163f8a3509fa3916482)
---
 lib/std/array_list.zig | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/lib/std/array_list.zig b/lib/std/array_list.zig
index 6ccb3c54e3e3..ab5985685c14 100644
--- a/lib/std/array_list.zig
+++ b/lib/std/array_list.zig
@@ -414,11 +414,14 @@ pub fn AlignedManaged(comptime T: type, comptime alignment: ?mem.Alignment) type
         pub fn shrinkRetainingCapacity(self: *Self, new_len: usize) void {
             assert(new_len <= self.items.len);
             self.items.len = new_len;
+            @memset(self.items[new_len..], undefined);
         }
 
+        /// Reduce length to 0.
         /// Invalidates all element pointers.
         pub fn clearRetainingCapacity(self: *Self) void {
             self.items.len = 0;
+            @memset(self.items, undefined);
         }
 
         /// Invalidates all element pointers.
-- 
2.52.0

